nautobot/dcim/models/devices.py

```diff
@@ -1375,3 +1386,329 @@ def clean(self):
+        on_delete=models.PROTECT,
+        related_name="modules",
+        blank=True,
+        null=True,
+    )
+    # TODO: add software support for Modules
+
+    clone_fields = [
+        "module_type",
+        "role",
+        "tenant",
+        "location",
+        "status",
+    ]
+
+    natural_key_field_names = ["location", "module_type", "parent_module_bay", "serial"]
+
+    class Meta:
+        ordering = ("parent_module_bay", "location", "module_type", "asset_tag", "serial")
+        constraints = [
+            # Database constraint to make the parent_model_bay and location fields mutually exclusive
+            models.CheckConstraint(
+                check=models.Q(parent_module_bay__isnull=False, location__isnull=True)
+                | models.Q(parent_module_bay__isnull=True, location__isnull=False),
+                name="dcim_module_parent_module_bay_xor_location",
+            ),
+            models.UniqueConstraint(
+                fields=["module_type", "serial"],
+                name="dcim_module_module_type_serial_unique",
+            ),
+        ]
+
+    def __str__(self):
+        serial = f" (Serial: {self.serial})" if self.serial else ""
+        asset_tag = f" (Asset Tag: {self.asset_tag})" if self.asset_tag else ""
+        display = str(self.module_type) + serial + asset_tag
+        if self.location:
+            return f"{display} at location {self.location}"
+        else:
+            return f"{display} installed in {self.parent_module_bay}"
+
+    @property
+    def device(self):
+        """Walk up parent chain to find the Device that this Module is installed in, if one exists."""
+        if self.parent_module_bay is None:
+            return None
+        return self.parent_module_bay.device
+
+    def clean(self):
+        super().clean()
+
+        # Validate that the Module is associated with a Location or a ModuleBay
+        if self.parent_module_bay is None and self.location is None:
+            raise ValidationError("One of location or parent_module_bay must be set")
+
+        # Validate location
+        if self.location is not None:
+            if self.parent_module_bay is not None:
+                raise ValidationError("Only one of location or parent_module_bay must be set")
+
+            if ContentType.objects.get_for_model(self) not in self.location.location_type.content_types.all():
+                raise ValidationError(
+                    {"location": f'Modules may not associate to locations of type "{self.location.location_type}".'}
+                )
+
+    def save(self, *args, **kwargs):
+        is_new = not self.present_in_database
+
+        if self.serial == "":
+            self.serial = None
+        if self.asset_tag == "":
+            self.asset_tag = None
+
+        super().save(*args, **kwargs)
+
+        # If this is a new Module, instantiate all related components per the ModuleType definition
+        if is_new:
+            self.create_components()
```

Comment:
Do I remember correctly that there's supposed to be some logic about letting component names have a token for the module-bay position that gets substituted in when the module gets installed into a specific bay? So the same line card would have `Ethernet1/1` if it's in bay 1 but `Ethernet2/1` if it's in bay 2? Fine with me if that's future work at this point.

Reply:
That's correct. I was planning on doing that after the UI is completed